### PR TITLE
BREAKING: Change `thread.id` type from integer to string

### LINKED ISSUE #546: [general, attributes] Make the `thread.id` attribute a string
Context: https://opentelemetry.io/docs/specs/semconv/general/attributes/#general-thread-attributes

In C++, there is no guaranteed integral conversion of `std::this_thread::get_id()` (https://github.com/googleapis/google-cloud-cpp/pull/13151/files#r1397504258)

So we are storing the attribute as a string in our implementation. Can we update the spec to make the attribute value a string or an int?

### PR DESCRIPTION
**Pull request by pyohannes**
Fixes #546 

## Changes

As discussed in #546, `thread.id` for managed runtimes isn't guaranteed to be an integer across all cases (C++ being the exception).

This change makes `thread.id` a string and thus makes it possible to set this value across all languages.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: docs/attributes-registry/thread.md [hunk 1]
```diff
@@ -10,6 +10,6 @@ These attributes may be used for any operation to store information about a thre
 <!-- semconv registry.thread(omit_requirement_level) -->
 | Attribute  | Type | Description  | Examples  |
 |---|---|---|---|
-| `thread.id` | int | Current "managed" thread ID (as opposed to OS thread ID). | `42` |
+| `thread.id` | string | Current "managed" thread ID (as opposed to OS thread ID). | `42` |
```
Comment (trisch-me) [inline docs/attributes-registry/thread.md:13]: There should be changes in yaml file as well, probably you forgot them to commit? @pyohannes

File: .chloggen/thread-id-type.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+change_type: breaking
+component: general
```
Comment (trisch-me) [inline .chloggen/thread-id-type.yaml:2]: I think component in this case should be `thread`

**Review by trisch-me — CHANGES_REQUESTED** [review]

File: docs/attributes-registry/thread.md (hunk 1 above)
Comment (pyohannes) [reply docs/attributes-registry/thread.md:13]: I committed those changes now.

**Review by trisch-me — APPROVED** [review]
lgtm

**Comment by pyohannes** [comment]
> I think we should advertise this breaking change widely before merging though, can you raise in both the semantic convention SIG meeting and the specification SIG meeting?

I can do that. I can also get up draft PRs for some languages.

**Comment by reyang** [comment]
Some runtimes might not use "thread" concept explicitly (e.g. coroutine) although from the OS level there is normally a thread ID.

Instead making a type change, consider having a primary (int) and secondary (string) ID.

In .NET there are two things:
* Managed Thread ID: https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.managedthreadid
* Lower level Thread ID (whatever provided by the operating system)

**Comment by pellared** [comment]
I think we should have a distinction between OS threads and "runtime" threads like .NET threads. Shouldn't the "runtime threads" semantic conventions be language specific? E.g. `jvm.thread.id` and `jvm.thread.name`. I think `thread.id` and `thread.name` be about OS threads.

**Comment by pyohannes** [comment]
> Instead making a type change, consider having a primary (int) and secondary (string) ID. In .NET there are two things:
>
> * Managed Thread ID: https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.managedthreadid
> * Lower level Thread ID (whatever provided by the operating system)

The idea was to cover the first (managed thread id) by `thread.id` and make it a string (see discussions in #546, in particular https://github.com/open-telemetry/semantic-conventions/issues/546#issuecomment-1978895370).

There's a separate issue for introducing an OS thread id (#812). This should be a number an map to ECS' [`process.thread.id`](https://www.elastic.co/guide/en/ecs/1.12/ecs-process.html#field-process-thread-id).

**Review by lalitb — APPROVED** [review]
Looks good from otel-cpp and otel-rust prospective. 
@open-telemetry/cpp-approvers @open-telemetry/rust-approvers

File: docs/attributes-registry/thread.md [hunk 3]
```diff
@@ -10,6 +10,6 @@ These attributes may be used for any operation to store information about a thre
 <!-- semconv registry.thread(omit_requirement_level) -->
```
Comment (tigrannajaryan) [inline docs/attributes-registry/thread.md:10]: Unrelated to the change in this PR: do we want to a "Document Status" for docs in the attributes-registry? I see we have them in the other docs but not here.

**Review by tigrannajaryan — APPROVED** [review]

**Comment by pellared** [comment]
This looks strange from Go perspective where we have goroutines, not threads. Setting `thread.id` for goroutine id would be awkward. 

Some references:

- https://pkg.go.dev/github.com/petermattis/goid
- https://pkg.go.dev/golang.org/x/exp/trace#GoID

Also e.g. for Python AFAIK there is something like [Tasks](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task).

**Comment by lalitb** [comment]
> This looks strange from Go perspective where we have goroutines, not threads. Setting `thread.id` for goroutine id would be awkward.
> 
> Some references:
> 
> * https://pkg.go.dev/github.com/petermattis/goid
> * https://pkg.go.dev/golang.org/x/exp/trace#GoID
> 
> Also e.g. for Python AFAIK there is something like [Tasks](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task).

I believe it would be similar challenge for most of the other languages if they choose to use async-runtime, even though they support threads.

**Comment by reyang** [comment]
> > Instead making a type change, consider having a primary (int) and secondary (string) ID. In .NET there are two things:
> > 
> > * Managed Thread ID: https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.managedthreadid
> > * Lower level Thread ID (whatever provided by the operating system)
> 
> The idea was to cover the first (managed thread id) by `thread.id` and make it a string (see discussions in #546, in particular [#546 (comment)](https://github.com/open-telemetry/semantic-conventions/issues/546#issuecomment-1978895370)).
> 
> There's a separate issue for introducing an OS thread id (#812). This should be a number and map to ECS' [`process.thread.id`](https://www.elastic.co/guide/en/ecs/1.12/ecs-process.html#field-process-thread-id).

I don't think a single "managed thread id (string type)" across all language runtimes would work.
Take .NET for example, there are three things [ManagedThreadId](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.managedthreadid) (which is an integer), [thread name](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.name?view=net-8.0) which is a string, and the native operating system level thread id.

I think @pellared pointed out the right direction https://github.com/open-telemetry/semantic-conventions/pull/836#issuecomment-2023142230. We should have the OS level thread id #812 and let each language runtime decide how to model their threading/concurrency (rather than introducing a common thread id across all runtimes).

**Review by reyang — CHANGES_REQUESTED** [review]
https://github.com/open-telemetry/semantic-conventions/pull/836#issuecomment-2023445936

**Comment by trisch-me** [comment]
> We should have the OS level thread id https://github.com/open-telemetry/semantic-conventions/issues/812 and let each language runtime decide how to model their threading/concurrency

currently we have defined in semconv `thread.id` as managed thread. If we go this road it means we should use this field for OS thread id and for managed define it per language?

**Comment by reyang** [comment]
> > We should have the OS level thread id #812 and let each language runtime decide how to model their threading/concurrency
> 
> @reyang currently we have defined in semconv `thread.id` as managed thread. If we go this road it means you propose to use this field for OS thread id and for managed define it per language?

Yes. One small correction "for managed, define it per language/**runtime** if it makes sense"

**Comment by pyohannes** [comment]
> Take .NET for example, there are three things [ManagedThreadId](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.managedthreadid) (which is an integer), [thread name](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.name?view=net-8.0) which is a string, and the native operating system level thread id (which is an integer).

As currently specified, `thread.id` covers the `ManagedThreadId` (managed in the examples for populating this attribute), `thread.name` already exists, and #812 is about introducing the OS thread id.

> Yes. One small correction "for managed, define it per language/runtime if it makes sense"

Currently, `thread.id` is explicitly defined to _not_ be the OS thread id, as the description says "Current managed thread ID (as opposed to OS thread ID)".

The [table with examples of how to obtain this id](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attributes.md#general-thread-attributes) cover 6 languages, some more (for example Rust) could be added. I think this is enough coverage to merit a standardized approach across those 6+ languages who have the concept of a managed thread id. The problem that triggered this PR is the fact that two of those languages (C++ and Rust) cannot express this thread id as an integer. Languages who don't have this concept (which might be the exceptions) could model this in a different language-specific way.

However, re-thinking this and also given that this is a breaking change, it might make sense to come up with a comprehensive and consistent approach that covers all of OS thread id, managed thread id, and thread name, and try to make a single breaking change instead of piecemeal changes. I'll bring it up in the next semconv WG meeting.

**Comment by breedx-splk** [comment]
@pyohannes any updates on this? At minimum, this needs a rebase. 👍🏻

**Comment by pyohannes** [comment]
@breedx-splk See my comment https://github.com/open-telemetry/semantic-conventions/issues/546#issuecomment-2055998851.

This will require more discussions, and as there exist a lot of instrumentations using `thread.id`, we'd only want to break this once we have a comprehensive solution for both OS thread id and name, as well as managed thread id and name. And we'd only want to break things once with a solution that we are confident can be declared stable.

With that I'll close this PR. We'll stay with the status quo until we this comprehensive proposal mentioned above.

**Pull request closed without merging**